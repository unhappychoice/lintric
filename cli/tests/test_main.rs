use insta::assert_snapshot;
use std::fs;
use std::path::PathBuf;
use std::process::Command;

#[test]
fn test_basic_analysis() {
    let temp_dir = PathBuf::from("tmp");
    fs::create_dir_all(&temp_dir).expect("Unable to create test directory");

    let temp_file_path = temp_dir.join("temp_test_file.rs");
    fs::write(&temp_file_path, "let a = 1;\nlet b = a + 1;\n").expect("Unable to write test file");

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("--verbose")
        .arg(temp_file_path.to_string_lossy().replace("\\", "/"))
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_multiple_files_analysis() {
    let temp_dir = PathBuf::from("tmp");
    fs::create_dir_all(&temp_dir).expect("Unable to create test directory");

    let file1_path = temp_dir.join("file1.rs");
    fs::write(&file1_path, "let x = 1;\nlet y = x + 1;\n").expect("Unable to write file1.rs");

    let file2_path = temp_dir.join("file2.ts");
    fs::write(&file2_path, "const a = 1;\nlet b = a + 1;\n").expect("Unable to write file2.ts");

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg(temp_dir.to_string_lossy().replace("\\", "/"))
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_json_output() {
    let fixture_path = "../core/tests/rust/fixtures/complex_rust_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("--json")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_complex_rust_analysis() {
    let fixture_path = "../core/tests/rust/fixtures/complex_rust_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("--verbose")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_complex_typescript_analysis() {
    let fixture_path = "../core/tests/typescript/fixtures/complex_typescript_code.ts";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("--verbose")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_debug_ast_rust() {
    let fixture_path = "../core/tests/rust/fixtures/complex_rust_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("debug")
        .arg("ast")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_debug_ast_typescript() {
    let fixture_path = "../core/tests/typescript/fixtures/complex_typescript_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("debug")
        .arg("ast")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_debug_definition_rust() {
    let fixture_path = "../core/tests/rust/fixtures/complex_rust_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("debug")
        .arg("definition")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_debug_definition_typescript() {
    let fixture_path = "../core/tests/typescript/fixtures/complex_typescript_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("debug")
        .arg("definition")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_debug_dependency_rust() {
    let fixture_path = "../core/tests/rust/fixtures/complex_rust_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("debug")
        .arg("dependency")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}

#[test]
fn test_debug_dependency_typescript() {
    let fixture_path = "../core/tests/typescript/fixtures/complex_typescript_code.rs";

    let output = Command::new("cargo")
        .arg("run")
        .arg("--package")
        .arg("lintric-cli")
        .arg("--")
        .arg("debug")
        .arg("dependency")
        .arg(fixture_path)
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let _stderr = String::from_utf8_lossy(&output.stderr);

    assert!(output.status.success());
    assert_snapshot!(stdout);
}
