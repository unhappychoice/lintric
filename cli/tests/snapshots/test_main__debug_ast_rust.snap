---
source: cli/tests/test_main.rs
expression: out
---
(source_file
  (struct_item
    name: (type_identifier "Point")
    body: (field_declaration_list
      (field_declaration
        name: (field_identifier "x")
        type: (primitive_type "i32")
      )
      (field_declaration
        name: (field_identifier "y")
        type: (primitive_type "i32")
      )
    )
  )
  (function_item
    name: (identifier "add")
    parameters: (parameters
      (parameter
        pattern: (identifier "a")
        type: (primitive_type "i32")
      )
      (parameter
        pattern: (identifier "b")
        type: (primitive_type "i32")
      )
    )
    return_type: (primitive_type "i32")
    body: (block
      (let_declaration
        pattern: (identifier "result")
        value: (binary_expression
          left: (identifier "a")
          right: (identifier "b")
        )
      )
      (identifier "result")
    )
  )
  (function_item
    name: (identifier "main")
    parameters: (parameters)
    body: (block
      (let_declaration
        pattern: (identifier "p1")
        value: (struct_expression
          name: (type_identifier "Point")
          body: (field_initializer_list
            (field_initializer
              field: (field_identifier "x")
              value: (integer_literal "1")
            )
            (field_initializer
              field: (field_identifier "y")
              value: (integer_literal "2")
            )
          )
        )
      )
      (let_declaration
        pattern: (identifier "p2")
        value: (struct_expression
          name: (type_identifier "Point")
          body: (field_initializer_list
            (field_initializer
              field: (field_identifier "x")
              value: (integer_literal "3")
            )
            (field_initializer
              field: (field_identifier "y")
              value: (integer_literal "4")
            )
          )
        )
      )
      (let_declaration
        pattern: (identifier "p3")
        value: (struct_expression
          name: (type_identifier "Point")
          body: (field_initializer_list
            (field_initializer
              field: (field_identifier "x")
              value: (call_expression
                function: (identifier "add")
                arguments: (arguments
                  (field_expression
                    value: (identifier "p1")
                    field: (field_identifier "x")
                  )
                  (field_expression
                    value: (identifier "p2")
                    field: (field_identifier "x")
                  )
                )
              )
            )
            (field_initializer
              field: (field_identifier "y")
              value: (call_expression
                function: (identifier "add")
                arguments: (arguments
                  (field_expression
                    value: (identifier "p1")
                    field: (field_identifier "y")
                  )
                  (field_expression
                    value: (identifier "p2")
                    field: (field_identifier "y")
                  )
                )
              )
            )
          )
        )
      )
      (let_declaration
        pattern: (identifier "p4")
        value: (block
          (let_declaration
            pattern: (identifier "p5")
            value: (struct_expression
              name: (type_identifier "Point")
              body: (field_initializer_list
                (field_initializer
                  field: (field_identifier "x")
                  value: (integer_literal "5")
                )
                (field_initializer
                  field: (field_identifier "y")
                  value: (integer_literal "6")
                )
              )
            )
          )
          (let_declaration
            pattern: (identifier "p6")
            value: (identifier "p5")
          )
          (identifier "p6")
        )
      )
      (let_declaration
        pattern: (identifier "x")
        value: (integer_literal "1")
      )
      (let_declaration
        pattern: (identifier "y")
        value: (binary_expression
          left: (identifier "x")
          right: (integer_literal "1")
        )
      )
      (let_declaration
        pattern: (identifier "z")
        value: (binary_expression
          left: (identifier "y")
          right: (identifier "x")
        )
      )
      (expression_statement
        (macro_invocation
          macro: (identifier "println")
          (token_tree
            (string_literal ""{:?}""
              (string_content)
            )
            (identifier "p3")
          )
        )
      )
      (expression_statement
        (macro_invocation
          macro: (identifier "println")
          (token_tree
            (string_literal ""{:?}""
              (string_content)
            )
            (identifier "p4")
          )
        )
      )
    )
  )
)
